src/rosdistro/manifest_provider/gitlab.py

```diff
@@ -0,0 +1,156 @@
+# Software License Agreement (BSD License)
+#
+# Copyright (c) 2021, Open Source Robotics Foundation, Inc.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above
+#    copyright notice, this list of conditions and the following
+#    disclaimer in the documentation and/or other materials provided
+#    with the distribution.
+#  * Neither the name of Open Source Robotics Foundation, Inc. nor
+#    the names of its contributors may be used to endorse or promote
+#    products derived from this software without specific prior
+#    written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import json
+import os
+import re
+from urllib.request import urlopen, Request
+from urllib.error import URLError
+from urllib.parse import quote as urlquote
+
+from catkin_pkg.package import parse_package_string
+
+from rosdistro.source_repository_cache import SourceRepositoryCache
+from rosdistro import logger
+
+GITLAB_PRIVATE_TOKEN = os.getenv('GITLAB_PRIVATE_TOKEN', None)
+
+def _gitlab_urlopen(url):
+    req = Request(url)
+    if GITLAB_PRIVATE_TOKEN:
+        req.add_header('Private-Token', GITLAB_PRIVATE_TOKEN)
+    logger.warn('Performing GitLab API query "%s"' % (url,))
+    return urlopen(req)
+
+
+def _gitlab_api_query(server, path, resource, attrs):
+    url = 'https://%s/api/v4/projects/%s/%s' % (server, urlquote(path, safe=''), resource)
+    if attrs:
+        url += '?' + '&'.join(k + '=' + urlquote(str(v), safe='') for k, v in attrs.items())
```

Comment:
Is there a reason not to use [`urllib.parse.urlencode`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode)?

```suggestion
        url += '?' urlencode(attrs)
```

Reply:
No reason at all. 1b46c7b